File: api/src/test/java/net/blueshell/api/factory/model/AnswerFactory.java
package net.blueshell.api.factory.model;

import net.blueshell.api.testutil.ModelTestUtils;
import com.github.javafaker.Faker;
import lombok.RequiredArgsConstructor;
import net.blueshell.api.common.enums.QuestionType;
import net.blueshell.api.model.survey.Answer;
import net.blueshell.api.model.survey.Question;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Factory for Answer model test instances.
 */
@Component
@RequiredArgsConstructor
public class AnswerFactory {

    private static final AtomicLong counter = new AtomicLong(1000);
    private final Faker faker;
    private final QuestionFactory questionFactory;

    public Answer createBasic() {
        Answer answer = new Answer();
        ModelTestUtils.setId(answer, generateId());

        Question q = questionFactory.createFull();
        answer.setQuestionId(q.getId());

        if (q.getType() == QuestionType.CHECKBOX || q.getType() == QuestionType.RADIO) {
            List<Boolean> selections = new ArrayList<>();
            if (q.getChoiceLabels() != null) {
                for (int i = 0; i < q.getChoiceLabels().size(); i++) {
                    selections.add(faker.bool().bool());
                }
            }
            answer.setOptionSelections(selections);
            answer.setTextResponse(null);
        } else {
            answer.setOptionSelections(null);
            answer.setTextResponse(faker.lorem().sentence());
        }
        return answer;
    }

    public Answer createFull() {
        return createBasic();
    }

    public Answer createWithCustomizations(java.util.function.Consumer<Answer> customizer) {
        Answer a = createFull();
        customizer.accept(a);
        return a;
    }

    private Long generateId() {
        return counter.incrementAndGet();
    }
}

File: api/src/test/kotlin/net/blueshell/api/factory/model/survey/AnswerFactory.kt
package net.blueshell.api.factory.model.survey

import com.github.javafaker.Faker
import net.blueshell.api.common.enums.QuestionType
import net.blueshell.api.model.survey.Answer
import org.springframework.stereotype.Component
import java.util.concurrent.atomic.AtomicLong
import java.util.function.Consumer

/**
 * Factory for Answer model test instances.
 */
@Component
class AnswerFactory(
    private val faker: Faker,
    private val questionFactory: QuestionFactory
) {

    fun createBasic(): Answer {
        val answer = Answer()
        val question = questionFactory.createFull()

        if (question.type == QuestionType.CHECKBOX || question.type == QuestionType.RADIO) {
            val selections = mutableListOf<Boolean>()
            val choices = question.choiceLabels
            if (choices != null) {
                repeat(choices.size) {
                    selections.add(faker.bool().bool())
                }
            }
            answer.optionSelections = selections
            answer.textResponse = null
        } else {
            answer.optionSelections = null
            answer.textResponse = faker.lorem().sentence()
        }
        return answer
    }

    fun createFull(): Answer = createBasic()

    fun createWithCustomizations(customizer: Consumer<Answer>): Answer {
        val answer = createFull()
        customizer.accept(answer)
        return answer
    }

    private fun generateId(): Long = COUNTER.incrementAndGet()

    private companion object {
        val COUNTER = AtomicLong(1000)
    }
}
